advisor/src/main/kotlin/advisors/NexusIq.kt

```diff
@@ -0,0 +1,169 @@
+import org.ossreviewtoolkit.utils.log
+import org.ossreviewtoolkit.utils.showStackTrace
+
+import retrofit2.Call
+
+/**
+ * A wrapper for [Nexus IQ Server](https://help.sonatype.com/iqserver) security vulnerability data.
+ */
+class NexusIq(
+    name: String,
+    config: AdvisorConfiguration
+) : Advisor(name, config) {
+    class Factory : AbstractAdvisorFactory<NexusIq>(ADVISOR_NAME) {
+        override fun create(config: AdvisorConfiguration) = NexusIq(advisorName, config)
+    }
+
+    companion object {
+        const val ADVISOR_NAME = "NexusIQ"
+    }
+
+    private val nexusIqConfig = config as NexusIqConfiguration
+
+    override suspend fun retrievePackageVulnerabilities(packages: List<Package>): Map<Package, List<AdvisorResult>> {
+        val serverUrl = nexusIqConfig.serverUrl
+        require(serverUrl != null) {
+            "No serverUrl configured for advisor '$ADVISOR_NAME'."
+        }
+
+        val service = NexusIqService.create(
+            serverUrl,
+            nexusIqConfig.username,
+            nexusIqConfig.password,
+            OkHttpClientHelper.buildClient()
+        )
+
+        val advisorDispatcher =
+            Executors.newSingleThreadExecutor(NamedThreadFactory(advisorName)).asCoroutineDispatcher()
+
+        return coroutineScope {
+            val startTime = Instant.now()
+
+            val components = packages.map { pkg ->
+                val packageUrl = buildString {
+                    append(pkg.purl)
+                    val purlType = pkg.id.getPurlType()
+                    if (purlType == Identifier.PurlType.MAVEN) append("?type=jar")
+                    if (purlType == Identifier.PurlType.PYPI) append("?extension=tar.gz")
+                }
+
+                NexusIqService.Component(packageUrl)
+            }
+
+            try {
+                val componentDetails =
+                    execute(withContext(advisorDispatcher) {
+                        service.getComponentDetails(NexusIqService.ComponentsWrapper(components))
+                    }).componentDetails
+                        .associateBy { it.component.packageUrl.substringBefore("?") }
+
+                val endTime = Instant.now()
+
+                packages.filter {
+                    componentDetails[it.id.toPurl()]?.securityData?.vulnerabilities?.isNotEmpty() ?: false
+                }.associateWith {
+                    val details = componentDetails[it.id.toPurl()]
+                    if (details != null) {
+                        val vulnerabilities = details.securityData.vulnerabilities
+                        listOf(
+                            AdvisorResult(
+                                vulnerabilities,
+                                AdvisorDetails(ADVISOR_NAME),
+                                AdvisorSummary(startTime, endTime)
+                            )
+                        )
+                    } else {
+                        emptyList()
+                    }
+                }
```

Comment:
I believe this could be replaced with

                packages.mapNotNullTo(mutableListOf()) { pkg ->
                    componentDetails[pkg.id.toPurl()]?.let { details ->
                        pkg to listOf(
                            AdvisorResult(
                                details.securityData.vulnerabilities,
                                AdvisorDetails(ADVISOR_NAME),
                                AdvisorSummary(startTime, endTime)
                            )
                        )
                    }
                }.toMap()

Reply:
Your suggested replacement would contain packages with empty vulnerabilities.
I combined your and my solution to get it working:
```
packages.filter {
    componentDetails[it.id.toPurl()]?.securityData?.vulnerabilities?.isNotEmpty() ?: false
}.mapNotNullTo(mutableListOf()) { pkg ->
    componentDetails[pkg.id.toPurl()]?.let { details ->
        pkg to listOf(
            AdvisorResult(
                details.securityData.vulnerabilities,
                AdvisorDetails(ADVISOR_NAME),
                AdvisorSummary(startTime, endTime)
            )
        )
    }
}.toMap()
```
Not sure if there is a nicer solution.

Reply:
I tried to find a nicer solution, see my new review.